### ISSUE TITLE
Feature Request: Other Apps

### ISSUE THREAD

**Issue by ghost** [quentinlesceller/macmediakeyforwarder#14]
I specifically would like this same functionality for PlexAmp, but I'm sure other people would like it for other apps.

**Comment by tboggia** [comment]
Yeah, I use Tidal and would love this to work.

**Comment by ghost** [comment]
+1 (For Roon)

**Comment by herewegoinvt** [comment]
> I specifically would like this same functionality for PlexAmp, but I'm sure other people would like it for other apps.

Second PlexAmp

**Comment by Offbeatmammal** [comment]
+1 for https://github.com/jpochyla/psst
rather than add/maintain a long list of music apps, is it possible to add a config option that includes the path to the app the user wants to launch and have that triggered when needed?

**Comment by joaofrgomes** [comment]
> Yeah, I use Tidal and would love this to work.

Same here! I've recently started using Tidal, and would love to have this as an option, instead of having to keep the “Pause if no other player is running” toggled and risking having the media player controls targeting Safari instead…

**Comment by draxx31** [comment]
> +1 for https://github.com/jpochyla/psst rather than add/maintain a long list of music apps, is it possible to add a config option that includes the path to the app the user wants to launch and have that triggered when needed?

I fully agree with this proposition. In fact, we need to be able to bypass "apple music" to interect with our preferred music app ( Deezer, Spotify, Tidal, ...). I do not know if @quentinlesceller is something possible, but getting the ability to choose the targeted app will be great like [MusicDecoy](https://github.com/FuzzyIdeas/MusicDecoy) does for example.

**Comment by ospfranco** [comment]
This is only possible if the apps expose commands via [osascript](https://victorscholz.medium.com/what-is-osascript-e48f11b8dec6) or [SBApplication](https://developer.apple.com/documentation/scriptingbridge/sbapplication) which Apple Music and Spotify do, otherwise it's not possible. For each one of the apps you guys mentioned it will probably not be possible as vendors are lazy with their software and do not ship this deep macOS integration support.

**Comment by muuvmuuv** [comment]
Yeah second this! What about PWA apps, like if I have Netflix installed via Safari/Chrome?

**Comment by quentinlesceller** [comment]
Thanks everyone for the requests. I looked into what it would take to support the apps mentioned here (TIDAL, PlexAmp, Deezer, Roon, Psst, browser PWAs like Netflix).

Mac Media Key Forwarder controls players through Scripting Bridge / AppleScript. It calls commands like `playpause`, `next track` and `previous track` on the target app. That only works for apps that ship an AppleScript scripting dictionary exposing those commands. Apple Music and Spotify both do, which is why they are the two currently supported.

I checked each requested app:

| Requested app | Type | AppleScript dictionary? | Supportable today? |
|---|---|---|---|
| TIDAL | Electron | No | No |
| PlexAmp | Electron | No | No |
| Deezer | Electron | No | No |
| Roon | proprietary | No | No |
| Psst | open source | No | No |
| Netflix PWA | browser | No | No |

None of them expose a scripting dictionary with playback commands, so there is nothing for the app to call. This is a limitation on the app vendors side, not something the forwarder can work around with the current approach.

Apps that do ship a dictionary (for example VLC and Swinsian) could be added with the existing mechanism, so if there is interest in those I can look at wiring them up.

**Comment by andrwmai** [comment]
Hello @quentinlesceller, would you mind taking a look at https://cider.sh/? It's an Apple Music client. Would love for MMKF to support it. 🙏🏻

**Comment by quentinlesceller** [comment]
Update: [Cider](https://cider.sh) is now supported as of [v4.1.0](https://github.com/quentinlesceller/macmediakeyforwarder/releases/tag/v4.1.0) (#39). Unlike the other apps discussed above, Cider exposes a local REST API, which made an integration possible without an AppleScript dictionary — thanks @andrwmai for requesting it and testing.

The situation for TIDAL, PlexAmp, Deezer, Roon, Psst and browser PWAs is unchanged: they expose neither a scripting dictionary nor a control API, so there is still nothing for the forwarder to call. If one of them ships either mechanism, I'm happy to look again.
